subiquity/ui/views/ubuntu_pro.py

```diff
@@ -81,23 +82,94 @@ def valid_char(self, ch: str) -> bool:
+        format. """
+        if len(self.token.value) not in range(24, 30):
+            return _("invalid token: should be between 24 and 30 characters")
+
+
+class UpgradeModeForm(Form):
+    """ Represents a form requesting the Ubuntu Pro credentials.
+    +---------------------------------------------------------+
+    | (X)  Add token manually                                 |
+    |      Token: C123456789ABCDEF                            |
+    |             This is your Ubuntu Pro token               |
+    |                                                         |
+    |                       [ Continue ]                      |
+    |                       [ Back     ]                      |
+    +---------------------------------------------------------+
     """
     cancel_label = _("Back")
+    ok_label = _("Continue")
+    group: List[RadioButtonField] = []
+
+    with_contract_token = RadioButtonField(
+            group, _("Add token manually"),
+            help=NO_HELP)
+    with_contract_token_subform = SubFormField(
+            ContractTokenForm, "", help=NO_HELP)
+
+    def __init__(self, initial) -> None:
+        """ Initializer that configures the callback to run when the radio is
+        checked/unchecked. Since there is a single radio for now, it can
+        only be unchecked programmatically. """
+        super().__init__(initial)
+        connect_signal(self.with_contract_token.widget,
+                       'change', self._toggle_contract_token_input)
+
+    def _toggle_contract_token_input(self, sender, new_value):
+        """ Enable/disable the sub-form that requests the contract token. """
+        self.with_contract_token_subform.enabled = new_value
+
+
+class UpgradeYesNoForm(Form):
+    """ Represents a form asking if we want to upgrade to Ubuntu Pro.
+    +---------------------------------------------------------+
+    | (X)  Upgrade to Ubuntu Pro                              |
+    |                                                         |
+    | ( )  Do this later                                      |
+    |      You can always enable Ubuntu Pro later via the     |
+    |      'ua attach' command                                |
+    |                                                         |
+    |                       [ Continue ]                      |
+    |                       [ Back     ]                      |
+    +---------------------------------------------------------+
+    """
 
-    UATokenField = simple_field(UATokenEditor)
+    cancel_label = _("Back")
+    ok_label = _("Continue")
+    group: List[RadioButtonField] = []
 
-    token = UATokenField(_("Ubuntu Pro token:"), help=ua_help)
+    upgrade = RadioButtonField(
+            group, _("Upgrade to Ubuntu Pro"),
+            help=NO_HELP)
+    skip = RadioButtonField(
+            group, _("Do this later"),
+            help=_("You can always enable Ubuntu Pro later via the"
```

Comment:
Check the indent on this line.  It's supposed to be down another line, and over.

Reply:
I moved this line down for now. Changing the indent seems challenging without leaning on dirty hacks.